File: 2023/day22/solution.py
#!/usr/bin/env python3
import sys
from dataclasses import dataclass
from typing import Self
import colorsys as coloursys
from functools import cache
from collections import defaultdict

# yapf: disable
sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True
INF = 10 ** 10

@dataclass
class Point():
  x: int; y: int; z: int
  def __mul__(self, val: int): return Point(self.x * val, self.y * val, self.z * val)
  def __add__(self, other: Self): return Point(self.x + other.x, self.y + other.y, self.z + other.z)
  def __sub__(self, other: Self): return Point(self.x - other.x, self.y - other.y, self.z - other.z)
  def __repr__(self): return f'({self.x},{self.y},{self.z})'
  def __getitem__(self, index: int): return (self.x, self.y, self.z)[index]
  def __lt__(self, other: Self): return (self.x, self.y, self.z) < (other.x, other.y, other.z)

@dataclass
class Brick():
  p0: Point; p1: Point; index:int
  def translate(self, val: Point): self.p0+=val; self.p1+=val
  #TODO: most of these properties should be precalculated really
  @property
  def w(self): return abs(self.p0.x-self.p1.x)+1
  @property
  def d(self): return abs(self.p0.y-self.p1.y)+1
  @property
  def h(self): return abs(self.p0.z-self.p1.z)+1
  @property
  def vol(self): return self.w * self.d * self.h
  @property
  def min(self): return Point(min(self.p0.x, self.p1.x), min(self.p0.y, self.p1.y), min(self.p0.z, self.p1.z))
  @property
  def max(self): return Point(max(self.p0.x, self.p1.x), max(self.p0.y, self.p1.y), max(self.p0.z, self.p1.z))
  @property
  def xy(self): return {(x,y) for x in range(self.min.x, self.max.x + 1) for y in range(self.min.y, self.max.y + 1)}
  @property
  def xz(self): return {(x,z) for x in range(self.min.x, self.max.x + 1) for z in range(self.min.z, self.max.z + 1)}
  @property
  def yz(self): return {(y,z) for y in range(self.min.y, self.max.y + 1) for z in range(self.min.z, self.max.z + 1)}
  @property
  def xyz(self): return {(x,y,z) for x in range(self.min.x, self.max.x + 1) for y in range(self.min.y, self.max.y + 1) for z in range(self.min.z, self.max.z + 1)}
  def __repr__(self): return f'<Brick index={self.index} p0={self.p0} p1={self.p1} w={self.w} d={self.d} h={self.h} xy={self.xy}>'
  def __hash__(self): return self.index
  @classmethod
  def from_str(cls, line: str, index:int): return cls(*[Point(*map(int,p.split(','))) for p in line.split('~')], index) #type: ignore

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i': FILENAME = 'input.txt'
      case '-p1': PART1, PART2 = True, False
      case '-p2': PART1, PART2 = False, True
      case '-p0': PART1, PART2 = False, False
      case _: raise Exception(f'unknown {arg=}')


#yapf: enable
def debug(level: int, *args, **kwargs):
  if level < DEBUG: print(*args, **kwargs)


def render(bricks: list[Brick], title: str = ''):
  # official aoc fancy drawing maker 3000
  if title: print(f'\x1b[1m{title}:\x1b[0m')
  min_x = min_y = min_z = INF
  max_x = max_y = max_z = -INF
  volume: dict[tuple[int, int, int], int] = {}
  for brick in bricks:
    min_x, min_y, min_z = min(min_x, brick.min.x), min(min_y, brick.min.y), min(min_z, brick.min.z)
    max_x, max_y, max_z = max(max_x, brick.max.x), max(max_y, brick.max.y), max(max_z, brick.max.z)
    for xyz in brick.xyz:
      volume[xyz] = brick.index
  min_z = 1
  # basic colours: https://www.shellhacks.com/bash-colors/
  # rgb: fg: 38;2;r;g;b; bg: 48;2;r;g;b
  bg_empty = '0;0;0'
  bg_brick = '255;255;255'
  fg_lightness = 0.2
  for y in range(min_y, max_y + 1):
    # zx slice, top to bottom
    for z in range(max_z, min_z - 1, -1):
      row = [f'\x1b[48;2;{bg_empty}m \x1b[0m'] * (max_x - min_x + 1)
      for x in range(min_x, max_x + 1):
        i = volume.get((x, y, z))
        if i is None: continue
        value = f'{i+10:X}'[-1]  #aoc label from A
        hue = 1.0 / len(bricks) * i  # not an off-by one since hue=0 and hue=1 are the same
        r, g, b = [int(x * 255) for x in coloursys.hls_to_rgb(hue, fg_lightness, 1.0)]
        coloured = f'\x1b[38;2;{r};{g};{b};1;48;2;{bg_brick}m{value}\x1b[0m'
        row[x] = coloured
      print(f'''{'  ' if title else ''}{y=:3d} {z=:3d}: {''.join(row)}''')
    print()


def get_bricks() -> list[Brick]:
  # do the drop in here since both parts require it
  with open(FILENAME, 'r') as f:
    bricks = [Brick.from_str(x, i) for i, x in enumerate(f.read().splitlines())]
  if DEBUG: render(bricks, 'initial')

  # sort on min.z so we drop in the right order
  bricks.sort(key=lambda x: x.min.z)

  # make a dict of xys and zs
  min_x = min_y = INF
  max_x = max_y = -INF
  for brick in bricks:
    min_x, min_y = min(min_x, brick.min.x), min(min_y, brick.min.y)
    max_x, max_y = max(max_x, brick.max.x), max(max_y, brick.max.y)
  z_belows = {(x, y): 0 for x in range(min_x, max_x + 1) for y in range(min_y, max_y + 1)}

  for above in bricks:
    # lookup the max z below our xys, translate to that+1, update dict
    z_below = max(z_belows[xy] for xy in above.xy)
    debug(1, f'{above=} {z_below=}')
    above.translate(Point(0, 0, z_below - above.min.z + 1))
    debug(2, f'  translated {above=}')
    z_belows |= {xy: above.max.z for xy in above.xy}

  # relative min.zs are now invalid
  if DEBUG: render(bricks, 'dropped')
  return bricks


def part1():
  # the question only wants to know which bricks could be *independently* removed
  # make a set of potential removals. loop over aboves. find their belows. if count belows==1, remove belows from potentials

  bricks = get_bricks()
  potentials = {*bricks}
  bricks.sort(key=lambda x: x.max.z, reverse=True)
  for i, above in enumerate(bricks):
    debug(1, f'{i=} {above=}')
    belows = set()
    for below in bricks[i + 1:]:
      if above.min.z - 1 > below.max.z: break  # max.z reverse sort lets us break out early
      if above.min.z - 1 == below.max.z and above.xy.intersection(below.xy):
        debug(2, f'  {below=}')
        belows.add(below)
    if len(belows) == 1:
      debug(1, f'  cannot remove {belows}')
      potentials.difference_update(belows)

  print(f'part 1: {len(potentials)}')
  # 519


def part2():
  # like a dependency tree, except all belows have to be removed for our item to be removed

  # build dicts of above and below index sets
  bricks = get_bricks()
  above_indices: dict[int, set[int]] = defaultdict(set)
  below_indices: dict[int, set[int]] = defaultdict(set)

  bricks.sort(key=lambda x: x.min.z)
  for i, below in enumerate(bricks):
    debug(1, f'{i=} {below=}')
    for above in bricks[i + 1:]:
      if below.max.z + 1 < above.min.z: break
      if below.max.z + 1 == above.min.z and below.xy.intersection(above.xy):
        debug(2, f'  {above=}')
        above_indices[below.index].add(above.index)

  bricks.sort(key=lambda x: x.max.z, reverse=True)
  for i, above in enumerate(bricks):
    debug(1, f'{i=} {above=}')
    for below in bricks[i + 1:]:
      if above.min.z - 1 > below.max.z: break
      if above.min.z - 1 == below.max.z and above.xy.intersection(below.xy):
        debug(2, f'  {below=}')
        below_indices[above.index].add(below.index)

  # surprisingly not too slow
  total = 0
  for brick in bricks:
    removed = {brick.index}
    found = True
    while found:
      found = False
      for index in removed.copy():
        for above_index in above_indices[index]:
          if above_index in removed: continue  #  this is a little faster than doing a difference
          if len(below_indices[above_index].difference(removed)) == 0:
            removed.add(above_index)
            found = True
    debug(0, f'{brick=} {len(removed)=}')
    total += len(removed) - 1

  print(f'part 2: {total}')
  #109531


if PART1: part1()
if PART2: part2()

File: 2023/day22/solution.ts
#!/usr/bin/env -S deno --allow-read
// #region base aoc template
const Maths = Math;

class Args {
  constructor(public filename = 'input.txt', public debug = 0, public part1 = true, public part2 = true) {
    for (const arg of Deno.args) {
      const [key, value] = [arg.slice(0, 2), (fallback = 0) => Number(arg.slice(2) || fallback)];
      if (key === '-d') debug = value(1);
      else if (key === '-e') filename = `example${arg.slice(2)}.txt`;
      else if (key === '-i') filename = 'input.txt';
      else if (key === '-p') [part1, part2] = [(value(0) & 1) === 1, (value(0) & 2) === 2];
      else throw new Error(`unrecognised arg="${arg}"`);
    }
    [this.filename, this.debug, this.part1, this.part2] = [filename, debug, part1, part2];
    console.log(`args: {filename: "${filename}", debug: ${debug}, part1: ${part1}, part2: ${part2} }`);
  }
}

class HashedSet<K, H> {
  //FIXME: send some time writing this properly and benchmarking
  private map: Map<H, K>;
  private hashFn: (key: K) => H;
  private hashSet: Set<H>;
  constructor(hashFn: (key: K) => H, iterable?: Iterable<K>, public readonly readOnly = false) {
    this.hashFn = hashFn;
    if (typeof iterable !== 'undefined') {
      this.map = new Map<H, K>(Array.from(iterable).map((item) => [this.hashFn(item), item]));
    } else {
      this.map = new Map<H, K>();
    }
    this.hashSet = new Set<H>(this.map.keys());
  }
  add = (key: K) => {
    if (this.readOnly) throw new Error('do not');
    const hash = this.hashFn(key);
    this.map.set(hash, key);
    this.hashSet.add(hash);
  };
  clear = () => {
    if (this.readOnly) throw new Error('do not');
    this.map.clear();
    this.hashSet.clear();
  };
  delete = (key: K) => {
    if (this.readOnly) throw new Error('do not');
    this.map.delete(this.hashFn(key));
    this.hashSet.clear();
  };
  has = (key: K) => this.map.has(this.hashFn(key));
  get size() {
    return this.map.size;
  }
  values = () => this.map.values();
  hashes = () => this.hashSet.values();

  //TODO: add more methods
  intersection = (other: HashedSet<K, H>) =>
    new HashedSet<K, H>(
      this.hashFn,
      this.hashSet
        .intersection(other.hashSet)
        .values()
        .map((hash) => this.map.get(hash)!)
    );
  intersects = (other: HashedSet<K, H>) => this.hashSet.intersection(other.hashSet).size > 0;
  [Symbol.iterator]() {
    return this.values();
  }
}

class HashedMap<K, V, H> {
  private map: Map<H, { key: K; value: V }>;
  constructor(public hashFn: (key: K) => H, iterable?: Iterable<readonly [K, V]>) {
    if (typeof iterable !== 'undefined') {
      this.map = new Map<H, { key: K; value: V }>(Array.from(iterable).map(([key, value]) => [this.hashFn(key), { key, value }]));
    } else {
      this.map = new Map<H, { key: K; value: V }>();
    }
  }
  clear = () => this.map.clear();
  delete = (key: K) => this.map.delete(this.hashFn(key));
  get = (key: K) => this.map.get(this.hashFn(key))?.value;
  has = (key: K) => this.map.has(this.hashFn(key));
  keys = () => this.map.values().map((v) => v.key);
  set = (key: K, value: V) => this.map.set(this.hashFn(key), { key, value });
  get size() {
    return this.map.size;
  }
  values = () => this.map.values().map((v) => v.value);
  entries = () => this.map.values().map((v) => [v.key, v.value]);
  [Symbol.iterator]() {
    return this.entries();
  }
}

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (args.debug >= level) console.debug(...data);
};

const args = new Args();
const input = await Deno.readTextFile(args.filename);
// #endregion

interface Vec3 {
  x: number;
  y: number;
  z: number;
}
interface Vec2 {
  x: number;
  y: number;
}
class Brick {
  private _c0: Vec3;
  private _c1: Vec3;
  public xy: HashedSet<Vec2, number>;
  constructor(public readonly id: number, c0: Vec3, c1: Vec3) {
    [this._c0, this._c1] = [c0, c1];
    const xys = new Array<Vec2>();
    for (let x = Maths.min(c0.x, c1.x); x <= Maths.max(c0.x, c1.x); ++x) {
      for (let y = Maths.min(c0.y, c1.y); y <= Maths.max(c0.y, c1.y); ++y) {
        xys.push({ x, y });
      }
    }
    this.xy = new HashedSet<Vec2, number>((key) => (key.x << 8) + key.y, xys, true);
  }
  get top(): number {
    return Maths.max(this._c0.z, this._c1.z);
  }
  get bottom(): number {
    return Maths.min(this._c0.z, this._c1.z);
  }
  moveZ = (value: number): void => {
    this._c0.z += value;
    this._c1.z += value;
  };
}

const parseInput = () =>
  input
    .matchAll(/^(\d+),(\d+),(\d+)~(\d+),(\d+),(\d+)$/gm)
    .map<Brick>((match, i) => {
      const [_, c0xStr, c0yStr, c0zStr, c1xStr, c1yStr, c1zStr] = match;
      return new Brick(
        i,
        { x: Number(c0xStr), y: Number(c0yStr), z: Number(c0zStr) },
        { x: Number(c1xStr), y: Number(c1yStr), z: Number(c1zStr) }
      );
    })
    .toArray();

// by val so we don't update the original... though maybe the array entries are still passed by ref idk
const dropBricks = (...bricks: Brick[]) => {
  const heights = new HashedMap<Vec2, number, number>((key) => (key.x << 8) + key.y);
  bricks.sort((a, b) => a.bottom - b.bottom);
  for (const brick of bricks) {
    const zBelow = brick.xy.values().reduce((acc, item) => Maths.max(heights.get(item) ?? 0, acc), 0);
    const moveBy = zBelow - brick.bottom + 1;
    debug(1, 'found', { brick, zBelow, moveBy });
    brick.moveZ(moveBy);
    debug(1, 'dropped', { brick });
    for (const xy of brick.xy) heights.set(xy, brick.top);
  }
  return bricks;
};

const part1 = () => {
  const bricks = dropBricks(...parseInput());

  const possibilities = new HashedSet<Brick, number>((key) => key.id, bricks);
  for (const above of bricks) {
    const belows = bricks.filter((below) => below.top === above.bottom - 1 && below.xy.intersects(above.xy));
    debug(1, { above, belows });
    // remove below from possibilities if there is exactly one
    if (belows.length === 1) belows.forEach((below) => possibilities.delete(below));
  }

  console.log('part 1:', possibilities.size);
};

const part2 = () => {
  const bricks = dropBricks(...parseInput());

  // refactor bricks into a more useful format
  interface BrickMap {
    aboves: Set<number>;
    belows: Set<number>;
  }
  const brickMap = new Map<number, BrickMap>(
    bricks.map((brick) => [
      brick.id,
      {
        aboves: new Set(
          bricks.filter((above) => above.bottom === brick.top + 1 && above.xy.intersects(brick.xy)).map((above) => above.id)
        ),
        belows: new Set(
          bricks.filter((below) => below.top === brick.bottom - 1 && below.xy.intersects(brick.xy)).map((below) => below.id)
        ),
      },
    ])
  );

  // then just keep looping over fallen and removing aboves with no non-fallen belows until we find no more
  let total = 0;
  for (const removed of brickMap.keys()) {
    const fallen = new Set<number>([removed]);
    let found = true;
    while (found) {
      found = false;
      for (const brick of fallen) {
        for (const above of brickMap.get(brick)!.aboves) {
          if (fallen.has(above)) continue;
          if (brickMap.get(above)!.belows.difference(fallen).size === 0) {
            fallen.add(above);
            found = true;
          }
        }
      }
    }
    total += fallen.size - 1; // don't include removed brick
  }

  console.log('part 2:', total);
};

if (args.part1) part1();
if (args.part2) part2();
